pre_commit/languages/download.py

```diff
@@ -0,0 +1,267 @@
+get_default_version = lang_base.basic_get_default_version
+run_hook = lang_base.basic_run_hook
+IO_BUF_SIZE = 4096
+
+
+def get_env_patch(target_dir: str) -> PatchesT:
+    return (
+        ('PATH', (target_dir, os.pathsep, Var('PATH'))),
+    )
+
+
+@contextlib.contextmanager
+def in_env(prefix: Prefix, version: str) -> Generator[None, None, None]:
+    envdir = lang_base.environment_dir(prefix, ENVIRONMENT_DIR, version)
+    with envcontext(get_env_patch(envdir)):
+        yield
+
+
+class Platform:
+    OS: Final[MappingProxyType[str, str]] = MappingProxyType({
+        'Linux': 'linux',
+        'Darwin': 'darwin',
+        'Windows': 'windows',
+        'DragonFly': 'dragonfly',
+        'FreeBSD': 'freebsd',
+    })
+    CPU: Final[MappingProxyType[str, str]] = MappingProxyType({
+        'aarch64': 'arm64',
+        'aarch64_be': 'arm64be',
+        'arm': 'arm',
+        'i386': '386',
+        'i686': '386',
+        'x86_64': 'amd64',
+        'AMD64': 'amd64',
+        'ppc': 'ppc',
+        'ppc64': 'ppc64',
+        'ppc64le': 'ppc64le',
+    })
+
+    def __init__(self, value: str) -> None:
+        self.__value = value
+        os, cpu = self.parts
+        if os not in Platform.OS.values():
+            raise ValueError(f'invalid operating system `{os}`')
+        if cpu not in Platform.CPU.values():
+            raise ValueError(f'invalid CPU `{cpu}`')
+
+    @property
+    def parts(self) -> tuple[str, str]:
+        first, second = self.value.split('/', 1)
+        return (first, second)
+
+    @property
+    def os(self) -> str:
+        os, _ = self.parts
+        return os
+
+    @property
+    def cpu(self) -> str:
+        _, cpu = self.parts
+        return cpu
+
+    @property
+    def value(self) -> str:
+        return self.__value
+
+    @staticmethod
+    def host() -> Platform:
+        os = Platform.OS[platform.system()]
+        cpu = Platform.CPU[platform.machine()]
+        return Platform(f'{os}/{cpu}')
+
+    def __str__(self) -> str:
+        return f'{self.os}/{self.cpu}'
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, Platform):
+            return NotImplemented
+        return other.os == self.os and other.cpu == self.cpu
+
+
+class ChecksumError(ValueError):
+    def __init__(self, sri: SRI, message: str = 'Mismatched checksum'):
+        self.__sri = sri
+        self.__message = message
+
+    @property
+    def sri(self) -> SRI:
+        return self.__sri
+
+    @property
+    def message(self) -> str:
+        return self.__message
+
+    def __str__(self) -> str:
+        return f'''{self.__message}:
+ - expected: {self.sri.checksum}
+ - actual  : {self.sri.digest}'''
+
+
+class SRI:
+    def __init__(self, value: str):
+        self.__value = value
+        self._algorithm, self._checksum = self.value.split('-', 1)
+        if self.algorithm not in hashlib.algorithms_available:
+            raise ValueError('Invalid hashing algorithm.')
```

Comment:
It is good practice to provide the following in error messages:

* *what* the error is
* *why* the error occured
* *how* can I fix it

```suggestion
            raise ValueError(f"`{self.algorithm}` is not available, choose one of: {','.join(hashlib.algorithms_available)}`")
```